### PR TITLE
dashboard: List app history (release and scale)

### PR DESCRIPTION
**Pull request by jvatic**
The next step is adding a revert button to each item.

### CODE REVIEW

**Comment by jvatic** [comment]
There's a few intermittently failing tests due to failed port allocation, e.g. `router: tcp sync error: error binding to port (check if another service is listening on 127.0.0.1:45000): listen tcp4 127.0.0.1:45000: bind: address already in use`.

**Comment by jvatic** [comment]
ping @titanous for review

**Comment by titanous** [comment]
This is great! I think the list should have similar styling to the source history and icons for each event type.

**Comment by jvatic** [comment]
Added icons and more detail to scale events (now shows the delta).

File: controller/formation.go [hunk 1]
```diff
@@ -74,6 +74,14 @@ func (r *FormationRepo) Add(f *ct.Formation) error {
 	if err != nil {
 		return err
 	}
+	scale := &ct.Scale{
+		Processes: f.Processes,
+		ReleaseID: f.ReleaseID,
+	}
+	prevFormation, _ := scanFormation(tx.QueryRow("SELECT app_id, release_id, processes, created_at, updated_at FROM formations WHERE app_id = $1 AND release_id = $2", f.AppID, f.ReleaseID))
```
Comment (jvatic) [inline controller/formation.go]: This should use `r.Get(appID, releaseID)`

File: controller/formation.go [hunk 2]
```diff
@@ -143,6 +151,13 @@ func (r *FormationRepo) Remove(appID, releaseID string) error {
 	if err != nil {
 		return err
 	}
+	scale := &ct.Scale{
+		ReleaseID: releaseID,
+	}
+	prevFormation, _ := scanFormation(tx.QueryRow("SELECT app_id, release_id, processes, created_at, updated_at FROM formations WHERE app_id = $1 AND release_id = $2", appID, releaseID))
```
Comment (jvatic) [inline controller/formation.go]: Ditto

File: controller/client/client.go [hunk 3]
```diff
@@ -548,10 +549,62 @@ func (c *Client) StreamEvents(opts StreamEventsOptions, output chan *ct.Event) (
...
+	SinceID     *int64
+	Count       int
+}
+
+func (c *Client) ListEvents(opts ListEventsOptions) ([]*ct.Event, error) {
+	var events []*ct.Event
+	path, err := url.Parse("/events")
+	if err != nil {
+		return nil, err
+	}
```
Comment (titanous) [inline controller/client/client.go]: This should never error.

**Comment by titanous** [comment]
- [ ] The UI should not show the scale events associated with deploying a new release (currently there is a down+up shown)
- [ ] The log should scroll to the top when a new event is added

**Comment by jvatic** [comment]
- [ ] Fix bug causing scale UI to disappear when saving ENV

**Comment by jvatic** [comment]
@titanous comments addressed

**Comment by titanous** [comment]
Looks like there is still an event that is not getting hidden:

1. Scale app to 3 instances.
1. Set environment variable.
1. Note that there is an extra `web: 3` event before the new release event and after the scale up event.

File: controller/types/types.go [hunk 4]
```diff
@@ -217,6 +217,17 @@ type Event struct {
 	CreatedAt  *time.Time      `json:"created_at,omitempty"`
 }
 
+type Scale struct {
+	PrevProcesses map[string]int `json:"prev_processes,omitempty"`
+	Processes     map[string]int `json:"processes"`
+	ReleaseID     string         `json:"release_id"`
```
Comment (jvatic) [inline controller/types/types.go]: The json key should be `release` to be consistent with other types.

Comment (jvatic) [inline controller/types/types.go]: `App.ReleaseID` should also be updated.

**Comment by jvatic** [comment]
Filtered out all initial scale events (no prev release).

File: controller/client/client.go [hunk 5]
```diff
@@ -548,8 +588,18 @@ func (c *Client) StreamEvents(opts StreamEventsOptions, output chan *ct.Event) (
...
 		q.Set("object_id", opts.ObjectID)
 	}
+	if opts.Count > 0 {
+		q.Set("count", strconv.Itoa(opts.Count))
+	}
 	path.RawQuery = q.Encode()
-	return c.ResumingStream("GET", path.String(), output)
+	h := make(http.Header)
+	h.Set("Accept", "application/json")
+	res, err := c.RawReq("GET", path.String(), h, nil, &events)
```
Comment (titanous) [inline controller/client/client.go]: Why `RawReq` over `Get`?

**Comment by titanous** [comment]
LGTM pending comment.

File: controller/client/client.go (hunk 5 above)
Comment (jvatic) [inline controller/client/client.go]: So the header can be set.

**Pull request merged**